File: ios/ExpoArcgisSceneView.swift
import ArcGIS
import Combine
import ExpoModulesCore
import SwiftUI

/// Holds the externally-provided ArcGIS `Scene` and bridges events to JS.
final class SceneViewModel: ObservableObject {
  // `ArcGIS.Scene` qualified to avoid collision with `SwiftUI.Scene` (this file imports SwiftUI).
  @Published private(set) var scene: ArcGIS.Scene?
  @Published private(set) var graphicsOverlays: [GraphicsOverlay] = []
  @Published private(set) var analysisOverlays: [AnalysisOverlay] = []
  @Published private(set) var camera: Camera?
  /// Bumped on each camera change so the SwiftUI `.task(id:)` re-runs and re-animates.
  @Published private(set) var cameraVersion = 0
  @Published private(set) var sunLighting: SceneView.SunLighting = .off
  @Published private(set) var atmosphereEffect: SceneView.AtmosphereEffect = .horizonOnly
  @Published private(set) var sunDate = Date(timeIntervalSince1970: 1_372_683_600)
  /// The view proxy captured from `SceneViewReader`, used for `identify` (not published).
  var proxy: SceneViewProxy?

  var onLoaded: (() -> Void)?
  var onLoadError: ((String) -> Void)?
  var onTap: ((_ latitude: Double, _ longitude: Double, _ screenX: Double, _ screenY: Double) -> Void)?

  func setScene(_ scene: ArcGIS.Scene?) {
    self.scene = scene
  }

  func setGraphicsOverlays(_ overlays: [GraphicsOverlay]) {
    graphicsOverlays = overlays
  }

  func setAnalysisOverlays(_ overlays: [AnalysisOverlay]) {
    analysisOverlays = overlays
  }

  func setCamera(_ camera: Camera) {
    self.camera = camera
    cameraVersion += 1
  }

  func setSunLighting(_ value: SceneView.SunLighting) { sunLighting = value }
  func setAtmosphereEffect(_ value: SceneView.AtmosphereEffect) { atmosphereEffect = value }
  func setSunDate(_ value: Date) { sunDate = value }
}

/// SwiftUI host for the ArcGIS `SceneView`. Loads the scene, reports the result, and forwards taps.
struct ExpoArcgisSceneContainer: View {
  @ObservedObject var model: SceneViewModel

  var body: some View {
    if let scene = model.scene {
      SceneViewReader { proxy in
        SceneView(
          scene: scene,
          graphicsOverlays: model.graphicsOverlays,
          analysisOverlays: model.analysisOverlays
        )
          // ArcGIS lighting modifiers return `SceneView`, so they precede the SwiftUI modifiers.
          .sunLighting(model.sunLighting)
          .atmosphereEffect(model.atmosphereEffect)
          .sunDate(model.sunDate)
          .onSingleTapGesture { screenPoint, scenePoint in
            // SceneView delivers an optional `Point` (a 3D tap can miss the globe).
            guard let scenePoint else { return }
            let wgs84 = GeometryEngine.project(scenePoint, into: .wgs84) ?? scenePoint
            model.onTap?(wgs84.y, wgs84.x, Double(screenPoint.x), Double(screenPoint.y))
          }
          .onAppear { model.proxy = proxy }
          .task(id: ObjectIdentifier(scene)) {
            do {
              try await scene.load()
              model.onLoaded?()
            } catch is CancellationError {
              // Superseded by a newer scene; ignore.
            } catch {
              model.onLoadError?(error.localizedDescription)
            }
          }
          .task(id: model.cameraVersion) {
            guard let camera = model.camera else { return }
            _ = await proxy.setViewpointCamera(camera, duration: 0.5)
          }
      }
    }
  }
}

/// Declarative 3D scene host. Renders the `SceneRef` passed as the `scene` view prop.
class ExpoArcgisSceneView: ExpoView {
  private let onSceneLoaded = EventDispatcher()
  private let onSceneLoadError = EventDispatcher()
  private let onTap = EventDispatcher()

  private let model = SceneViewModel()
  private var hostingController: UIHostingController<ExpoArcgisSceneContainer>?

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    clipsToBounds = true

    model.onLoaded = { [weak self] in
      self?.onSceneLoaded(["spatialReferenceWkid": NSNull()])
    }
    model.onLoadError = { [weak self] message in
      self?.onSceneLoadError(["message": message])
    }
    model.onTap = { [weak self] latitude, longitude, screenX, screenY in
      self?.onTap([
        "mapPoint": ["latitude": latitude, "longitude": longitude],
        "screenPoint": ["x": screenX, "y": screenY],
      ])
    }

    let hostingController = UIHostingController(rootView: ExpoArcgisSceneContainer(model: model))
    hostingController.view.backgroundColor = .clear
    hostingController.view.frame = bounds
    hostingController.view.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    addSubview(hostingController.view)
    self.hostingController = hostingController
  }

  /// Receives the native scene (by reference) from the `<Scene>` SharedObject.
  func setScene(_ ref: SceneRef?) {
    model.setScene(ref?.scene)
    // The scene may be replaced asynchronously (e.g. once a mobile scene package loads) — re-render.
    ref?.onSceneChanged = { [weak self] scene in self?.model.setScene(scene) }
  }

  /// Retries loading the scene (Loadable pattern) — useful after a network outage. Re-emits the result.
  func retryLoad() async throws {
    guard let scene = model.scene else { return }
    do {
      try await scene.retryLoad()
      onSceneLoaded(["spatialReferenceWkid": NSNull()])
    } catch {
      onSceneLoadError(["message": error.localizedDescription])
    }
  }

  /// Identifies the features under a screen point (3D). Mirrors `MapView.identify`.
  func identify(_ screenPoint: [String: Any], _ options: [String: Any]?) async throws -> [[String: Any]] {
    guard let proxy = model.proxy else { return [] }
    let point = CGPoint(
      x: (screenPoint["x"] as? NSNumber)?.doubleValue ?? 0,
      y: (screenPoint["y"] as? NSNumber)?.doubleValue ?? 0
    )
    let tolerance = (options?["tolerance"] as? NSNumber)?.doubleValue ?? 12
    let maxResults = (options?["maxResults"] as? NSNumber)?.intValue ?? 1
    let results = try await proxy.identifyLayers(
      screenPoint: point, tolerance: tolerance, maximumResultsPerLayer: maxResults
    )
    return results.map(serializeIdentifyResult)
  }

  /// Identifies popups under a screen point — evaluates each and returns `{ title, fields }`.
  func identifyPopups(_ screenPoint: [String: Any], _ options: [String: Any]?) async throws -> [[String: Any]] {
    guard let proxy = model.proxy else { return [] }
    let point = CGPoint(
      x: (screenPoint["x"] as? NSNumber)?.doubleValue ?? 0,
      y: (screenPoint["y"] as? NSNumber)?.doubleValue ?? 0
    )
    let tolerance = (options?["tolerance"] as? NSNumber)?.doubleValue ?? 12
    let maxResults = (options?["maxResults"] as? NSNumber)?.intValue ?? 1
    let results = try await proxy.identifyLayers(
      screenPoint: point, tolerance: tolerance, maximumResultsPerLayer: maxResults
    )
    return await serializePopups(results)
  }

  func setGraphicsOverlays(_ refs: [GraphicsOverlayRef]) {
    model.setGraphicsOverlays(refs.map { $0.overlay })
  }

  func setAnalysisOverlays(_ refs: [AnalysisOverlayRef]) {
    model.setAnalysisOverlays(refs.map { $0.overlay })
  }

  /// Animates the view to a runtime camera sent from JS.
  func setCamera(_ c: [String: Any]?) {
    guard let c, let position = c["position"] as? [String: Any] else { return }
    let point = scenePoint(position)
    let camera = Camera(
      location: point,
      heading: (c["heading"] as? NSNumber)?.doubleValue ?? 0,
      pitch: (c["pitch"] as? NSNumber)?.doubleValue ?? 0,
      roll: (c["roll"] as? NSNumber)?.doubleValue ?? 0
    )
    model.setCamera(camera)
  }

  func setSunLighting(_ s: String?) { model.setSunLighting(sunLightingMode(s)) }
  func setAtmosphereEffect(_ s: String?) { model.setAtmosphereEffect(atmosphereEffectMode(s)) }
  func setSunTime(_ ms: Double?) {
    if let ms { model.setSunDate(Date(timeIntervalSince1970: ms / 1000)) }
  }
}

/// Maps the JS sun-lighting union to the native `SceneView.SunLighting`.
func sunLightingMode(_ s: String?) -> SceneView.SunLighting {
  switch s {
  case "light": return .light
  case "lightAndShadows": return .lightAndShadows
  default: return .off
  }
}

/// Maps the JS atmosphere union to the native `SceneView.AtmosphereEffect`.
func atmosphereEffectMode(_ s: String?) -> SceneView.AtmosphereEffect {
  switch s {
  case "off": return .off
  case "realistic": return .realistic
  default: return .horizonOnly
  }
}

File: android/src/main/java/expo/modules/arcgis/ExpoArcgisSceneView.kt
package expo.modules.arcgis

import android.content.Context
import android.view.ViewGroup
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.findViewTreeLifecycleOwner
import com.arcgismaps.geometry.GeometryEngine
import com.arcgismaps.geometry.Point
import com.arcgismaps.geometry.SpatialReference
import com.arcgismaps.mapping.TimeExtent
import com.arcgismaps.mapping.view.AtmosphereEffect
import java.time.Instant
import com.arcgismaps.mapping.view.Camera
import com.arcgismaps.mapping.view.GlobeCameraController
import com.arcgismaps.mapping.view.LightingMode
import com.arcgismaps.mapping.view.OrbitLocationCameraController
import com.arcgismaps.mapping.view.SceneView
import com.arcgismaps.mapping.view.ScreenCoordinate
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.Promise
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

/** Declarative 3D scene host. Renders the [SceneRef] passed as the `scene` view prop. */
class ExpoArcgisSceneView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  private val onSceneLoaded by EventDispatcher<MapLoadedEventPayload>()
  private val onSceneLoadError by EventDispatcher<MapLoadErrorEventPayload>()
  private val onTap by EventDispatcher<TapEventPayload>()

  private val sceneView = SceneView(context).also {
    it.layoutParams = ViewGroup.LayoutParams(
      ViewGroup.LayoutParams.MATCH_PARENT,
      ViewGroup.LayoutParams.MATCH_PARENT
    )
    addView(it)
  }

  private val scope = CoroutineScope(Dispatchers.Main.immediate + SupervisorJob())
  private var loadJob: Job? = null
  private var observedLifecycle: Lifecycle? = null

  init {
    scope.launch {
      sceneView.onSingleTapConfirmed.collect { event ->
        val wgs84 = event.mapPoint?.let {
          GeometryEngine.projectOrNull(it, SpatialReference.wgs84()) as? Point
        }
        onTap(
          TapEventPayload(
            mapPoint = PointRecord(wgs84?.y ?: 0.0, wgs84?.x ?: 0.0),
            screenPoint = ScreenPointRecord(event.screenCoordinate.x, event.screenCoordinate.y)
          )
        )
      }
    }
  }

  /** Receives the native scene (by reference) from the `<Scene>` SharedObject. */
  fun setScene(ref: SceneRef?) {
    ref ?: return
    applyScene(ref.scene)
    // The scene may be replaced asynchronously (e.g. once a mobile scene package loads) — re-apply.
    ref.onSceneChanged = { newScene -> applyScene(newScene) }
  }

  private fun applyScene(scene: com.arcgismaps.mapping.ArcGISScene) {
    sceneView.scene = scene
    loadJob?.cancel()
    loadJob = scope.launch {
      scene.load()
        .onSuccess { onSceneLoaded(MapLoadedEventPayload()) }
        .onFailure { error ->
          onSceneLoadError(MapLoadErrorEventPayload(error.message ?: "Failed to load scene"))
        }
    }
  }

  /** Receives the graphics overlays declared as `<GraphicsOverlay>` children of the `<SceneView>`. */
  fun setGraphicsOverlays(refs: List<GraphicsOverlayRef>) {
    sceneView.graphicsOverlays.clear()
    sceneView.graphicsOverlays.addAll(refs.map { it.overlay })
  }

  /** Receives the analysis overlays declared as `<AnalysisOverlay>` children of the `<SceneView>`. */
  fun setAnalysisOverlays(refs: List<AnalysisOverlayRef>) {
    sceneView.analysisOverlays.clear()
    sceneView.analysisOverlays.addAll(refs.map { it.overlay })
  }

  /** Identifies the features under a screen point (3D). Mirrors `MapView.identify`. */
  fun identify(screenPoint: Map<String, Any?>, options: Map<String, Any?>?, promise: Promise) {
    val x = (screenPoint["x"] as? Number)?.toDouble() ?: 0.0
    val y = (screenPoint["y"] as? Number)?.toDouble() ?: 0.0
    val tolerance = (options?.get("tolerance") as? Number)?.toDouble() ?: 12.0
    val maxResults = (options?.get("maxResults") as? Number)?.toInt() ?: 1
    scope.launch {
      sceneView.identifyLayers(ScreenCoordinate(x, y), tolerance, false, maxResults)
        .onSuccess { results -> promise.resolve(results.map { serializeIdentifyResult(it) }) }
        .onFailure { promise.reject("IDENTIFY_ERROR", it.message ?: "Identify failed", it) }
    }
  }

  /** Identifies popups under a screen point — evaluates each and returns `{ title, fields }`. */
  fun identifyPopups(screenPoint: Map<String, Any?>, options: Map<String, Any?>?, promise: Promise) {
    val x = (screenPoint["x"] as? Number)?.toDouble() ?: 0.0
    val y = (screenPoint["y"] as? Number)?.toDouble() ?: 0.0
    val tolerance = (options?.get("tolerance") as? Number)?.toDouble() ?: 12.0
    val maxResults = (options?.get("maxResults") as? Number)?.toInt() ?: 1
    scope.launch {
      try {
        val results = sceneView.identifyLayers(ScreenCoordinate(x, y), tolerance, false, maxResults).getOrThrow()
        promise.resolve(serializePopups(results))
      } catch (e: Exception) {
        promise.reject("IDENTIFY_ERROR", e.message ?: "Identify failed", e)
      }
    }
  }

  /** Retries loading the scene (Loadable pattern) — useful after a network outage. Re-emits the result. */
  fun retryLoad(promise: Promise) {
    val scene = sceneView.scene ?: run { promise.resolve(null); return }
    scope.launch {
      scene.retryLoad()
        .onSuccess { onSceneLoaded(MapLoadedEventPayload()); promise.resolve(null) }
        .onFailure { error ->
          onSceneLoadError(MapLoadErrorEventPayload(error.message ?: "Failed to load scene"))
          promise.resolve(null)
        }
    }
  }

  /** Returns the terrain elevation (meters) at a point on the scene's base surface, or null. */
  fun getElevation(point: Map<String, Any?>, promise: Promise) {
    val scene = sceneView.scene ?: run { promise.resolve(null); return }
    val p = geometryFromDict(point) as? Point ?: run { promise.resolve(null); return }
    scope.launch {
      scene.baseSurface.getElevation(p)
        .onSuccess { promise.resolve(it) }
        .onFailure { e -> promise.reject("ELEVATION_ERROR", e.message ?: "Elevation query failed", e) }
    }
  }

  /** Animates the view to a runtime camera sent from JS. */
  fun setCamera(c: Map<String, Any?>?) {
    val position = c?.get("position") as? Map<*, *> ?: return
    val x = (position["x"] as? Number)?.toDouble() ?: 0.0
    val y = (position["y"] as? Number)?.toDouble() ?: 0.0
    val z = (position["z"] as? Number)?.toDouble()
    val point = if (z != null) Point(x, y, z, SpatialReference.wgs84())
    else Point(x, y, SpatialReference.wgs84())
    val camera = Camera(
      point,
      (c["heading"] as? Number)?.toDouble() ?: 0.0,
      (c["pitch"] as? Number)?.toDouble() ?: 0.0,
      (c["roll"] as? Number)?.toDouble() ?: 0.0,
    )
    scope.launch { sceneView.setViewpointCameraAnimated(camera, 0.5f) }
  }

  /** Sets or clears the scene's camera controller (orbit/globe). `null` restores the SDK default. */
  private var cameraControllerConfig: Map<String, Any?>? = null
  private var orbitGraphic: GraphicRef? = null

  fun setCameraController(c: Map<String, Any?>?) {
    cameraControllerConfig = c
    rebuildCameraController()
  }

  /** Stores the target graphic for an `orbitGeoElement` camera controller and rebuilds. */
  fun setOrbitGraphic(ref: GraphicRef?) {
    orbitGraphic = ref
    rebuildCameraController()
  }

  private fun rebuildCameraController() {
    val c = cameraControllerConfig
    sceneView.cameraController = when (c?.get("type") as? String) {
      "orbitLocation" -> {
        val target = c["target"] as? Map<*, *>
        val x = (target?.get("x") as? Number)?.toDouble() ?: 0.0
        val y = (target?.get("y") as? Number)?.toDouble() ?: 0.0
        val z = (target?.get("z") as? Number)?.toDouble()
        val point = if (z != null) Point(x, y, z, SpatialReference.wgs84())
                    else Point(x, y, SpatialReference.wgs84())
        val distance = (c["distance"] as? Number)?.toDouble() ?: 1500.0
        OrbitLocationCameraController(point, distance)
      }
      "orbitGeoElement" -> {
        val graphic = orbitGraphic
        if (graphic != null) {
          val distance = (c["distance"] as? Number)?.toDouble() ?: 1500.0
          com.arcgismaps.mapping.view.OrbitGeoElementCameraController(graphic.graphic, distance)
        } else GlobeCameraController()
      }
      "globe" -> GlobeCameraController()
      else -> GlobeCameraController() // null/absent → restore SDK default (GlobeCameraController)
    }
  }

  /** Sets the coordinate grid overlay from JS (null hides it). */
  fun setGrid(config: Map<String, Any?>?) {
    sceneView.grid = buildGrid(config)
      ?: com.arcgismaps.mapping.view.LatitudeLongitudeGrid().apply { isVisible = false }
  }

  /** Sun lighting mode (shadows). */
  fun setSunLighting(s: String?) {
    sceneView.sunLighting = when (s) {
      "light" -> LightingMode.Light
      "lightAndShadows" -> LightingMode.LightAndShadows
      else -> LightingMode.NoLight
    }
  }

  /** Atmosphere rendering. */
  fun setAtmosphereEffect(s: String?) {
    sceneView.atmosphereEffect = when (s) {
      "off" -> AtmosphereEffect.None
      "realistic" -> AtmosphereEffect.Realistic
      else -> AtmosphereEffect.HorizonOnly
    }
  }

  /** Sun position, as epoch milliseconds (affects shadow direction). */
  fun setSunTime(ms: Double?) {
    ms ?: return
    sceneView.sunTime = java.time.Instant.ofEpochMilli(ms.toLong())
  }

  /** Filters time-aware layers to a time window from JS (null shows all time steps). */
  fun setTimeExtent(config: Map<String, Any?>?) {
    if (config == null) { sceneView.setTimeExtent(null); return }
    val startMs = (config["startTime"] as? Number)?.toLong() ?: return
    val endMs = (config["endTime"] as? Number)?.toLong() ?: return
    sceneView.setTimeExtent(TimeExtent(Instant.ofEpochMilli(startMs), Instant.ofEpochMilli(endMs)))
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    val lifecycle = findViewTreeLifecycleOwner()?.lifecycle
    if (lifecycle != null && lifecycle !== observedLifecycle) {
      observedLifecycle?.removeObserver(sceneView)
      lifecycle.addObserver(sceneView)
      observedLifecycle = lifecycle
    }
  }

  override fun onDetachedFromWindow() {
    super.onDetachedFromWindow()
    loadJob?.cancel()
    observedLifecycle?.removeObserver(sceneView)
    observedLifecycle = null
  }
}
